compiler/rustc_hir_pretty/src/lib.rs

```diff
@@ -660,6 +660,28 @@ impl<'a> State<'a> {
                 }
                 self.bclose(item.span);
             }
+            hir::ItemKind::Macro(ref macro_def) => {
+                let (kw, has_bang) = if macro_def.macro_rules {
+                    ("macro_rules", true)
+                } else {
+                    self.print_visibility(&item.vis);
+                    ("macro", false)
+                };
+
+                self.print_mac_common(
+                    Some(MacHeader::Keyword(kw)),
+                    has_bang,
+                    Some(item.ident),
+                    macro_def.body.delim(),
+                    &macro_def.body.inner_tokens(),
+                    true,
+                    item.span,
+                );
+
+                if macro_def.body.need_semicolon() {
+                    self.word(";");
+                }
```

Comment:
Instead of copy-pasting this can be factored out into a method in `compiler\rustc_ast_pretty\src\pprust\state.rs`.

Reply:
It can, but it's a bit less trivial than it may at first appear. In `rustc_ast_pretty`, the visibility is an `ast::Visibility`. By contrast, in `rustc_hir_pretty`, the visibility is an `hir::Visibility`. As far as I can tell, there's no way to interconvert between the two, though I'd be overjoyed if someone told me I'm wrong about that.

That means that we'd end up needing to make the function generic with respect to the the type of visibility. The first pass I did has a function signature that looks like this:

```rust
fn print_mac_def<V>(
        &mut self,
        macro_def: &ast::MacroDef,
        ident: &Ident,
        vis: &V,
        sp: &Span,
        print_visibility: impl FnOnce(&mut Self, &V),
    )
```

It would probably be cleaner to use traits somehow, though I'm not sure what the best way to go about that is. In any case, this approach does add some amount of complexity. The DRY principle may be enough to justify that though.